Add monitor documentation (#1372)
hashgraph/hedera-mirror-node: docs/monitor.md

# Monitor

The monitor verifies end-to-end functionality of the Hedera network and generates metrics from the results. It supports
both publishing transactions to HAPI and subscribing to the mirror node API. Configuration is flexible and declarative
allowing one to express a mixture of transactions, their expected rate of publish and the tool will do its best to make
it so. By default, the monitor is already setup with a basic scenario that creates a topic, submits a message every 10

> seconds to it and verifies the messages is received via the mirror node's gRPC and REST APIs.

Author: Updated

> seconds to it and verifies the messages is received via the mirror node's gRPC and REST APIs.

Author: Updated

## Configuration

This section covers the higher level properties to consider before diving into the specific publish and subscribe
properties in the next sections. For a full list of configuration options see the [config](/docs/configuration.md)
documentation.

> First, make sure you are configured to talk to the correct Hedera network by setting `hedera.mirror.monitor.network` to

Author: Updated

> `MAINNET`, `PREVIEWNET` or `TESTNET`. If it's not one of these public environments, the network can be set to `OTHER`
> and `hedera.mirror.monitor.nodes` and `hedera.mirror.monitor.mirrorNode` properties should be filled in.

Reviewer: ```suggestion
`MAINNET`, `PREVIEWNET` or `TESTNET`. If you are not using one of these public environments, the network can be set to `OTHER`,
and `hedera.mirror.monitor.nodes` and `hedera.mirror.monitor.mirrorNode` properties should be filled in.
```

Author: Updated

Additionally, the operator information in `hedera.mirror.monitor.operator` is required and needs to be populated with a
valid payer account ID and its private key. Ensure this account has the necessary funds to publish transactions at the
rate you desire. The following is an example with a custom network configured:

```yaml
hedera:
  mirror:
    monitor:
      mirrorNode:
        grpc:
          host: 34.69.90.179
          port: 5600
        rest:
          host: 35.239.255.55
          port: 80
      network: OTHER
      nodes:
        - accountId: 0.0.3
          host: 34.94.106.61
        - accountId: 0.0.4
          host: 35.237.119.55
        - accountId: 0.0.5
          host: 35.245.27.193
        - accountId: 0.0.6
          host: 34.83.112.116
      operator:
        accountId: 0.0.1234
        privateKey: 104...
```

### Publish

> The monitor can be configured to publish multiple transaction types concurrently. It does this be specifying a list of

Author: Updated

publisher will round-robin this list of connections to determine which node to send to, ensuring the load is distributed
evenly across all nodes.

The `type` property specifies which transaction type to publish. It also affects which `properties` need to be
specified, with different transaction types requiring different properties to be set. See the
[TransactionType](/hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/TransactionType.java)
enum for a list of possible values for `type`. The properties can be seen as fields on the various
[TransactionSupplier](/hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier) classes that
the `TransactionType` enum references. Most of these properties have a default and don't need to be explicitly
specified, but some are empty and may need to be populated.

For example, if you want to publish a topic message, you would open the `TransactionType` class,
find `CONSENSUS_SUBMIT_MESSAGE`, then open the
[ConsensusSubmitMessageTransactionSupplier](/hedera-mirror-datagenerator/src/main/java/com/hedera/datagenerator/sdk/supplier/consensus/ConsensusSubmitMessageTransactionSupplier.java)

> class that it references. From there, you can see that fields `maxTransactionFee`, `message`, `retry`

Reviewer: ```suggestion
class that it references. From there, you can see that fields `maxTransactionFee`, `message`, `retry`,
```

Author: Updated

and `topicId` are available as properties. Only `topicId` doesn't have a default and will be required. Here's a YAML
excerpt that specifies some of those properties:

```yaml
publish:
  scenarios:
    - name: HCS Pinger
      properties:
        maxTransactionFee: 1000000
        message: Hello world!
        retry: true
        topicId: 0.0.1000
      record: 1.0
      tps: 0.1
      type: CONSENSUS_SUBMIT_MESSAGE
```

### Expression Syntax

> The monitor can automatically create account, token or topic entities on application startup using a special expression

Reviewer: ```suggestion
The monitor can automatically create account, token, and topic entities on application startup using a special expression
```

Author: Updated

syntax. This is useful to avoid boilerplate configuration and manual entity creation steps that vary per environment.
The syntax can currently only be used in `hedera.mirror.monitor.publish.scenarios.properties`
and `hedera.mirror.monitor.subscribe.grpc.topicId`.

> The syntax takes the form of `${type.name}` where `type` is one of `account`, `token` or `topic` and `name` is a

Reviewer: ```suggestion
The syntax takes the form of `${type.name}` where `type` is one of `account`, `token`, or `topic`, and `name` is a
```

Author: Updated

descriptive label. Based upon the entity type, it will create the appropriate entity on the network with default values.
The name label allows the same entity to be referenced in multiple places but only created once.

The following example uses the expression syntax to create the sender and recipient accounts as well as a token. These
same entities are created once and reused in both the token associate and the token transfer transaction.

```yaml
scenarios:
  - name: HTS associate
    limit: 1
    properties:
      accountId: ${account.them}
      tokenId: ${token.foobar}
    tps: 1
    type: TOKEN_ASSOCIATE
  - name: HTS transfer
    properties:
      recipientAccountId: ${account.them}
      senderAccountId: ${account.me}
      tokenId: ${token.foobar}
      transferType: BOTH
    tps: 1
    type: CRYPTO_TRANSFER
```

### Subscribe

> The monitor can optionally subscribe to either the mirror node gRPC or REST APIs. Each subscription type can have one or

Reviewer: It can do both at the same time if desired, right?

Author: Updated to make that clear

> more scenarios. For the REST API, it can verify a percentage of individual transactions have made it to the mirror node

Reviewer: ```suggestion
more scenarios. For the REST API, it can verify that a percentage of individual transactions have made it to the mirror node
```

Author: Updated
